src/lavinmq/amqp/exchange/topic.cr

```diff
@@ -2,8 +2,100 @@ require "./exchange"
 
 module LavinMQ
   module AMQP
+    class TopicBindingKey
+      abstract class Segment
+        abstract def match?(rk : RkPart) : Bool
+      end
+
+      class RkPart
+        getter :value, :next
+
+        def self.from(rk : Array(String)) : RkPart?
+          rk.reverse_each.reduce(nil) do |prev, v|
+            RkPart.new(v, prev)
+          end
+        end
+
+        def initialize(@value : String, @next : RkPart?)
+        end
+      end
+
+      class HashSegment < Segment
+        def initialize(@next : Segment?)
+        end
+
+        def match?(rk : RkPart) : Bool
+          if n = @next
+            loop do
+              return true if n.match?(rk)
+              rk = rk.next
+              break unless rk
+            end
+            return false
+          end
+          true
+        end
+      end
+
+      class StarSegment < Segment
+        def initialize(@next : Segment?)
+        end
+
+        def match?(rk : RkPart) : Bool
+          if check = @next
+            if n = rk.next
+              check.match?(n)
+            else
+              false
+            end
+          else
+            rk.next.nil?
+          end
+        end
+      end
+
+      class StringSegment < StarSegment
```

Comment:
Any reason a `StringSegment` is a `StarSegment` and not a `Segment`?

Reply:
Yes, `StringSegment::match?` uses the same logic as `StarSegment` it ju also needs to check that the string is equal. It can inherit directly from `Segment` but then we need to duplicate the logic form StarSegment. Not a big problem as I dont expect it to change very often.

Reply:
I agree that the inherit doesn't make much sense. I see no real problem with duplicate that small code snippet and have StringSegment inherit Segment directly. 

A risk with extracting it as a method in `Segment` is that it might expand for more use cases later.